### PR TITLE
raftstore: write tombstone state with writes in one batch when catching up logs for merge [release-2.1]

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
There are two bugs of #4581 and #4560 in master which fixed by https://github.com/tikv/tikv/pull/4595. But release-2.1 doesn't suffer from #4581, and to make changes as less as possible, this PR just write tombstone state with writes in one batch when catching up logs for merge instead of catching up logs through raft.

## What are the type of the changes? (mandatory)
- Bug fix (change which fixes an issue)

## How has this PR been tested? (mandatory)
unit-test

### CODE REVIEW

**Comment by Connor1996** [comment]
still have some compile errors, will be fixed later

**Comment by Connor1996** [comment]
The change is a little large which is not stable enough to merge into the release branch, so finding an easy way to work around.

**Comment by Connor1996** [comment]
PTAL @overvenus @BusyJay

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1701,22 +1701,30 @@ impl Peer {
...
             let cmd: RaftCmdRequest =
                 util::parse_data_at(entry.get_data(), entry.get_index(), &self.tag);
-            if !cmd.has_admin_request() {
-                continue;
+            if cmd.has_admin_request() {
+                let cmd_type = cmd.get_admin_request().get_cmd_type();
+                match cmd_type {
+                    AdminCmdType::TransferLeader | AdminCmdType::InvalidAdmin => {}
+                    _ => {
+                        // Any command that can change epoch or log gap should be rejected.
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why return errors for `VerifyHash`?

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -378,6 +382,7 @@ impl<'a> ApplyContextCore<'a> {
         if let Some(region) = stash.region {
             self.cbs.push(ApplyCallback::new(region));
         }
+        self.for_merge_source = false;
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: What if merge cascadingly?

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -2324,7 +2341,7 @@ mod tests {
             let key = format!("key_{}", i);
             wb.put(key.as_bytes(), b"value").unwrap();
         }
-        assert_eq!(should_write_to_engine(&req, wb.count()), false);
+        assert_eq!(should_write_to_engine(&req, wb.count(), false), false);
     }
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: Could you add a new case which always returns false even if wb > 4mb ?

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -2099,13 +2107,22 @@ impl Runner {
             {
                 let mut ctx = ApplyContext::new(&mut core, &mut self.delegates);
                 delegate.handle_raft_committed_entries(&mut ctx, apply.entries);
+                fail_point!(
+                    "skip_merge_tombstone_persist",
+                    { delegate.id() == 3 && delegate.region_id() == 1 },
+                    |_| {
+                        ctx.delegates.remove(&region_id);
+                        ctx.cbs.drain(..);
+                        ()
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: This line can be removed.

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -459,7 +465,7 @@ pub fn notify_stale_req(term: u64, cb: Callback) {
 }
 
 /// Check if a write is needed to be issued before handle the command.
-fn should_write_to_engine(cmd: &RaftCmdRequest, wb_keys: usize) -> bool {
+fn should_write_to_engine(cmd: &RaftCmdRequest, wb_keys: usize, for_merge_source: bool) -> bool {
     if cmd.has_admin_request() {
         match cmd.get_admin_request().get_cmd_type() {
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: Should we guard this check with `for_merge_source`?

Comment (Connor1996) [reply src/raftstore/store/worker/apply.rs]: no, when handle committed entries for merge, it is still possible to handle some admin requests(not in log gap, just not applied in time) that should write to engine.

**Comment by Connor1996** [comment]
PTAL again @BusyJay @overvenus

**Pull request merged**